### PR TITLE
Add minimum gas price

### PR DESCRIPTION
**Pull request by bowenwang1996**
Add minimum gas price to genesis config so that when there are not enough transactions, gas price will not become zero.

### CODE REVIEW

**Comment by bowenwang1996** [comment]
Division by zero will not happen given how we compute the new gas price. I don't understand at which point we switch to use initial gas price or previous gas price.

**Comment by nearmax** [comment]
> Division by zero will not happen given how we compute the new gas price.

I bad, what I meant is when there is not enough transactions we should fall back to either the gas price defined in the genesis, or the last price that was computed.

> I don't understand at which point we switch to use initial gas price or previous gas price.

We would have to figure out the exact mechanics here and make sure this is not abusable. Adding one more parameter to our config obscures the problem, but does not solve it -- how do we determine the minimum gas price? What if it is too low, can it stall some shard by making it non-profitable to validate? Can anything bad happen if it is too hight? Also, when we have enough transactions to calculate the gas cost, can it drop below the minimum gas cost defined in this config?

**Comment by bowenwang1996** [comment]
Gas price is computed every block, so not sure what you mean by "have enough transactions to compute gas cost". It seems that you are suggesting a dynamic approach to computing the minimum gas price, but some threshold has to be set (could be absolute or relative). I am also fine if we just use initial gas price as minimum gas price, but then we just need to take that into consideration when we set its value.

**Comment by nearmax** [comment]
Okay, I understand now our miscommunication. What you are saying is that when there are few transactions the gas price might drop too much, correct? 

I actually expect the completely opposite thing to happen when there are few transactions. **The blockchain without transactions is more expensive to run which will be reflected by the gas price going up.** Here is the explanation:

Gas price is only relevant in the context of the fees. There are accountable and non-accountable expenses for the validators. Accountable expenses are fees. Non-accountable expenses are more complex things that we do not account for with our fees, e.g. cost of serializing and deserializing chunk/header, cost of catching up with the network, cost of helping someone catch up with the network. The cost of these non-accountable expenses is amortized across the participating nodes, which is one of the major contributors to the gas price. The more transactions we have per block the easier it is to pay for the non-accountable expenses with little gas increase. Now imagine we start receiving fewer transactions per block, actually, let's imagine we receive 1 transaction per block. The fees from this single transaction is going to be the only source of reward (not accounting for the inflation) to the validators, which means the gas price will need to go 10x or 100x or 1000x up to make this single transaction in each block to pay for not accountable expenses.

The side effect of gas price going up when there are fewer transactions is that the blockchain has some gas price threshold after each it enters the death spiral -- fewer people use it due to high prices, which leads to prices going even higher up.

**Comment by bowenwang1996** [comment]
If we try to balance it this way, validators might have incentives to accept fewer transactions while still getting roughly the same amount of reward from executing transactions. Also the validator reward does not only come from transactions fees. When there are not enough transactions, they get reward mostly from inflation.

**Comment by nearmax** [comment]
> If we try to balance it this way, validators might have incentives to accept fewer transactions while still getting roughly the same amount of reward from executing transactions.

It is hypothetical whether validators will do that and whether this is going to be enough to avoid the death spiral.

> Also the validator reward does not only come from transactions fees. When there are not enough transactions, they get reward mostly from inflation.

So, hypothetically, inflation reward will stop the gas price going up into a death spiral. Do we have a formal way to analyse at what gas price it will stop growing up?

**Comment by bowenwang1996** [comment]
We unfortunately don't have any formal analysis on anything economics related, but I tend to believe that gas price should go up when transaction volume goes up so that validators get more reward when there are more transactions. It also makes sense that when there is more demand, the price is higher and vice versa.

**Comment by nearmax** [comment]
@evgenykuzyakov just reminded me that the gas price is not directly influenced by the validators  and we actually implement it to go up or down depending on whether the load of the blockchain goes up or down. @bowenwang1996 AFAIR you have implemented it, do you remember whether we have implemented the gas price to go up or to go down depending on the load? Do you remember where in the code we have this logic?

**Comment by bowenwang1996** [comment]
Yes the gas price will currently go to zero if there are not enough transactions. https://github.com/nearprotocol/nearcore/blob/f93e13c0be52b8069a98de4792b861f07787f003/core/primitives/src/block.rs#L424

**Comment by nearmax** [comment]
> Yes gas price will currently go to zero if there are not enough transactions.
> 
> https://github.com/nearprotocol/nearcore/blob/f93e13c0be52b8069a98de4792b861f07787f003/core/primitives/src/block.rs#L424

I think there is a problem with the linked code that will cause gas_used to almost always go down independently on how much the blockchain is used. `gas_limit` there refers to the gas limit of the block. In all cases, except for one exception listed below `gas_used <= gas_limit`, which means `numerator <= denominator`, which means `gas_price <= prev_gas_price`, meaning it will always go down or stay the same.

The only exception comes from the fact that when there is a large load we can accidentally use more gas than the gas limit, by a tiny bit amount, which is also going to be an almost random amount. And since it will be: tiny, accidental, and random; it will not be a value we can rely on to correctly push the price of the gas upwards.

**Comment by bowenwang1996** [comment]
Gas price will go up when `gas_used >= gas_limit/2`. I just checked again and it looks fine to me.

**Comment by nearmax** [comment]
> Gas price will go up when `gas_used >= gas_limit/2`. I just checked again and it looks fine to me.

So we have decided that we want our blocks to be half-full on average. This seems like a random constant that might be adjusted in the future (e.g. to make sure we utilize 80% of our costs). Should we make a parameter from it?

**Comment by bowenwang1996** [comment]
Not sure whether it's a random constant @ilblackdragon.

File: chain/chain/src/chain.rs [hunk 1]
```diff
@@ -152,6 +158,7 @@ pub struct ChainGenesis {
     pub time: DateTime<Utc>,
     pub gas_limit: Gas,
     pub gas_price: Balance,
+    pub min_gas_price: Balance,
```
Comment (ilblackdragon) [inline chain/chain/src/chain.rs]: should we have a field of type `BlockEconomicsConfig` here?

File: near/src/config.rs [hunk 2]
```diff
@@ -86,6 +86,9 @@ pub const INITIAL_GAS_LIMIT: Gas = 10_000_000;
 /// Initial gas price.
 pub const INITIAL_GAS_PRICE: Balance = 100;
```
Comment (ilblackdragon) [inline near/src/config.rs]: let's remove initial gas price. and for genesis use min_gas_price as the gas_price

**Comment by ilblackdragon** [comment]
@nearmax @bowenwang1996 
This constant is connected with constant of how long we also waiting for prev block before considering to skip it. E.g if you decide for blocks to be 80% full, you are expected to wait for prev block only for 20% more time then expected block time and which can be affected by other lags.

50% balances this, with min_block_production_delay is 1s (https://github.com/nearprotocol/nearcore/blob/master/near/src/config.rs#L52) and max_block_production_delay is 2s (https://github.com/nearprotocol/nearcore/blob/master/near/src/config.rs#L55).

Even though we can adjust all the coefficients together to a different value, 50% both seems natural and also matches suggestions by Vitalik and few other folks working on similar model.

**Comment by nearmax** [comment]
@ilblackdragon FYI, on a different, but relevant note. In the economics paper it says:

> We suggest that every block producers vote on `gasLimit` by providing a new value within +/-0.1% if the previous block has `gasUsed > 0.9 * gasLimit`.

This will almost never happen and if it will this event will be random. Our `gasLimit` is hard, meaning we define on the protocol, that as we process transactions/receipts as soon as we hit `gasLimit` with the last transaction/receipt we stop processing them. So in order to go 10% over the `gasLimit` we need to be lucky with getting this last transaction to be large enough (the order of transactions is defined in the protocol, so it will be hard for block producer to arrange that).

We need a replacement for the mechanism that determines when to vote on `gasLimit`.

**Comment by bowenwang1996** [comment]
I am confused as to why it is almost impossible to have `gasUsed > 0.9 * gasLimit` . I assume `gasLimit` will be a reasonably large value and transactions usually do not consume a ton of gas so it seems entirely plausible to have `gasUsed > 0.9 * gasLimit` if there are a lot of transactions.

**Comment by nearmax** [comment]
> I am confused as to why it is almost impossible to have `gasUsed > 0.9 * gasLimit` . I assume `gasLimit` will be a reasonably large value and transactions usually do not consume a ton of gas so it seems entirely plausible to have `gasUsed > 0.9 * gasLimit` if there are a lot of transactions.

Oh yeah, you are right. I misunderstood.

**Pull request merged**